plugins/processors/tag_limit/tag_limit.go

```diff
@@ -0,0 +1,71 @@
+package taglimit
+
+import (
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/processors"
+	"log"
+)
+
+const sampleConfig = `
+  ## Maximum number of tags to preserve
+  limit = 10
+
+  ## List of tags to preferentially preserve
+  keep = ["foo", "bar", "baz"]
+`
+
+type TagLimit struct {
+	Limit int      `toml:"limit"`
+	Keep  []string `toml:"keep"`
+}
+
+func (d *TagLimit) SampleConfig() string {
+	return sampleConfig
+}
+
+func (d *TagLimit) Description() string {
+	return "Restricts the number of tags that can pass through this filter and chooses which tags to preserve when over the limit."
+}
+
+func (d *TagLimit) Apply(in ...telegraf.Metric) []telegraf.Metric {
+	// convert list of tags to a map so we can do constant-time lookups
+	keepTags := make(map[string]string)
+	for _, tag := range d.Keep {
+		keepTags[tag] = ""
+	}
+	for _, point := range in {
+		pointTags := point.Tags()
```

Comment:
This function is pretty slow since it makes a copy of all the tags into a map.  Let's iterate over the TagList() and build up an array of tags we want to remove, then call RemoveTag on all of them.  We can initialize the list to the length of limit, and optionally, reuse the array each call to Apply.

Reply:
I don't think we can assume that the number of tags to drop is <= the limit number?  Or that the list of tags to drop would be consistent from metric to metric?

Anyway, sorry, I hadn't quite realized that point.Tags() was so expensive but you're quite right -- converting over to using TagList().